massa-execution-worker/src/execution.rs

```diff
@@ -960,15 +885,120 @@ impl ExecutionState {
+                slot, self.final_cursor
+            );
+        }
+
+        // if the slot was already executed, truncate active history to cancel the slot and all the ones after
+        if &self.active_cursor >= slot {
+            debug!(
+                "execute_candidate_slot: truncating down from slot {}",
+                self.active_cursor
+            );
+            self.active_history
+                .write()
+                .truncate_from(slot, self.config.thread_count);
+            self.active_cursor = slot
+                .get_prev_slot(self.config.thread_count)
+                .expect("overflow when iterating on slots");
+        }
+
+        let exec_out = self.execute_slot(slot, exec_target, selector);
+        debug!("execute_candidate_slot: execution finished");
+
+        // apply execution output to active state
+        self.apply_active_execution_output(exec_out);
+        debug!("execute_candidate_slot: execution state applied");
+    }
+
+    /// Execute an SCE-final slot
+    #[allow(clippy::borrowed_box)]
+    pub fn execute_final_slot(
+        &mut self,
+        slot: &Slot,
+        exec_target: &Option<(BlockId, Storage)>,
+        selector: &Box<dyn SelectorController>,
+    ) {
+        let target_id = exec_target.as_ref().map(|(b_id, _)| *b_id);
+        debug!(
+            "execute_final_slot: executing slot={} target={:?}",
+            slot, target_id
+        );
+
+        if slot <= &self.final_cursor {
+            debug!(
+                "execute_final_slot: final slot already executed (final_cursor = {})",
+                self.final_cursor
+            );
+            return;
+        }
+
+        // check if the final slot execution result is already cached at the front of the speculative execution history
+        let first_exec_output = self.active_history.write().0.pop_front();
+        if let Some(exec_out) = first_exec_output {
+            if &exec_out.slot == slot && exec_out.block_id == target_id {
+                // speculative execution front result matches what we want to compute
+
+                // apply the cached output and return
+                self.apply_final_execution_output(exec_out);
+
+                debug!("execute_final_slot: found in cache, applied cache");
+                return;
+            } else {
+                // speculative cache mismatch
+                warn!(
+                    "speculative execution cache mismatch (final slot={}/block={:?}, front speculative slot={}/block={:?}). Resetting the cache.",
+                    slot, target_id, exec_out.slot, exec_out.block_id
+                );
+            }
+        } else {
+            // cache entry absent
+            info!(
+                "speculative execution cache empty, executing final slot={}/block={:?}",
+                slot, target_id
+            );
+        }
+
+        // truncate the whole execution queue
+        self.active_history.write().0.clear();
```

Comment:
Why do you clear whole speculative cache when a block became final ?

Reply:
In most cases, the slot that is becoming final has already been executed as speculative in the past and is therefore present at the front of the active history.

So when a slot becomes final, we first check first element of the active history:
* if it is present and its input matches the one of the slot being finalized (same slot, same block ID or miss) then we can just remove that element from the front of the active history and apply its already-computed output to the final state. That's the normal situation
* if it is absent (history empty) it means that we just bootstrapped or that the execution is lagging a lot => we need to execute the slot now and apply its output to the final state
* if it is present but mismatches the slot being finalized, it means that the whole blockclique has changed suddenly (deep fork). This is theoretically not impossible but the probability of it happening is at the very limit of our security conditions. Hence the warning. In that case we need to discard the whole history and execute the slot being finialized, then apply its result to the final state.

In the code, we return early in the 1st case here https://github.com/massalabs/massa/blob/c0865fe4d79bb0f0e63c26332608b023a7697ebf/massa-execution-worker/src/execution.rs#L969-L972 and the clear doesn't happen.

In the 2 other cases the function continues and we do the history clear

Reply:
> and is therefore present at the front of the active history.

This is what my question is about why the final slot is at the front ? Because they become final one by one with the condition that the precedent one is final also ?

Reply:
Exactly, SCE finality is slot-sequential, so the next slot to finalize is almost certainly the earliest current speculative slot.